### PR TITLE
Fix vm group deletion

### PR DESCRIPTION
**Pull request by mmidolesov2**
Removed check if the vm to delete is the last one in the group and
replace it with poll cluster group values so we can check if there
is any instance left in the current vm group. If no instances left
in the group delete the group.

### CODE REVIEW

File: nova/virt/vmwareapi/vmops.py [hunk 1]
```diff
@@ -1497,19 +1498,16 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
                             continue
-
-                        for vm in group.vm:
-                            if vm.value == vm_ref.value and len(group.vm) == 1:
-                                cluster_util.delete_vm_group(
-                                            self._session, cluster.obj,
-                                            config_info_ex.group[key])
-                                break
-                        break
+                        else:
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: You've got a `continue` right above.You don't need the `else` and the extra indent.

File: nova/virt/vmwareapi/vmops.py [hunk 2]
```diff
@@ -1497,19 +1498,16 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
+                            for vm in group.vm:
+                                if vm.value == vm_ref.value:
+                                    vm_group = group
+                                    break
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Couldn't there be multiple groups in theory? If an admin added a VM to a group via vCenter or something?

File: nova/virt/vmwareapi/vmops.py [hunk 3]
```diff
@@ -1535,6 +1533,18 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
+                    cluster_config = self._session._call_method(
+                        vutil, "get_object_property", self._cluster,
+                        "configurationEx")
+                    for group in cluster_config.group:
+                        if group.name == vm_group.name:
+                            if not hasattr(group, 'vm'):
+                                cluster_util.delete_vm_group(
+                                    self._session, cluster.obj,
+                                    group)
+                                break
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Going with the logic above, the `break` should be done one indent less, i.e. when the group matching the name is found, not when the group did not have an attribute `vm`.

File: nova/virt/vmwareapi/vmops.py [hunk 4]
```diff
@@ -1535,6 +1533,18 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
                 self._session._call_method(self._session.vim,
                                            "UnregisterVM", vm_ref)
                 LOG.debug("Unregistered the VM", instance=instance)
+                if vm_group:
+                    cluster_config = self._session._call_method(
+                        vutil, "get_object_property", self._cluster,
+                        "configurationEx")
+                    for group in cluster_config.group:
+                        if group.name == vm_group.name:
+                            if not hasattr(group, 'vm'):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Are you sure that the group will not have the attribute anymore and that it's not an empty list?

File: nova/virt/vmwareapi/vmops.py [hunk 5]
```diff
@@ -1572,6 +1582,7 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
             vm_util.vm_ref_cache_delete(instance.uuid)
 
     def destroy(self, context, instance, destroy_disks=True):
+
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Please remove that newline.

File: nova/virt/vmwareapi/vmops.py (hunk 4 above)
Comment (mmidolesov2) [reply nova/virt/vmwareapi/vmops.py]: Yes - double checked it. when the last vm is removed is not [] but removed property instead.

Comment (joker-official) [reply nova/virt/vmwareapi/vmops.py]: Ok. Now it's triple-checked :smile:

File: nova/virt/vmwareapi/vmops.py [hunk 6]
```diff
@@ -1535,6 +1531,22 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
                                            "UnregisterVM", vm_ref)
                 LOG.debug("Unregistered the VM", instance=instance)
+                cluster_config = self._session._call_method(
+                    vutil, "get_object_property", self._cluster,
+                    "configurationEx")
+
+                # Check if the vm to unregister is the last vm in it's group
+                for group in cluster_config.group:
+                    for vm_group in vm_groups:
+                        if group.name == vm_group.name:
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: If all you need are the names of the groups the VM is in, then you should create a set of names and just do a
```python
if group.name not in vm_group_names:
    continue
```

File: nova/virt/vmwareapi/vmops.py [hunk 7]
```diff
@@ -1535,6 +1531,22 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
+                            with lockutils.lock(server_group_info.uuid,
+                                  lock_file_prefix='nova-vmware-server-group'):
+                                if not hasattr(group, 'vm'):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Wouldn't you have to lock earlier for this to be precise? You queried the groups without the lock above, so someone could have changed them in the meantime.

File: nova/virt/vmwareapi/vmops.py [hunk 8]
```diff
@@ -1535,6 +1530,20 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
                 self._session._call_method(self._session.vim,
                                            "UnregisterVM", vm_ref)
                 LOG.debug("Unregistered the VM", instance=instance)
+
+                with lockutils.lock(server_group_info.uuid,
+                                lock_file_prefix='nova-vmware-server-group'):
+                    cluster_config = vm_util.get_cluster_property(
+                        self._session, "configurationEx", self._cluster)
+                    # Check if the unregistered vm is the last vm in it's group
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: I think it's `its`, isn't it?

File: nova/virt/vmwareapi/cluster_util.py [hunk 9]
```diff
@@ -131,7 +131,7 @@ def update_placement(session, cluster, vm_ref, group_info):
     """Updates cluster for vm placement using DRS"""
     cluster_config = session._call_method(
         vutil, "get_object_property", cluster, "configurationEx")
-
+    config_spec = None
```
Comment (joker-official) [inline nova/virt/vmwareapi/cluster_util.py]: Why is this necessary?

File: nova/virt/vmwareapi/vm_util.py [hunk 10]
```diff
@@ -704,6 +705,13 @@ def _get_device_disk_type(device):
             return constants.DEFAULT_DISK_TYPE
 
 
+def get_cluster_property(session, prop_list, cluster):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vm_util.py]: Why is this cluster utility function in `vm_util` not `cluster_util`? Is it necessary to introduce a function for property-retrieval of a cluster, if you only need it once? I see other code just having that one line of code directly.

File: nova/virt/vmwareapi/vmops.py [hunk 11]
```diff
@@ -1491,25 +1491,20 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
-                                break
-                        break
+            if hasattr(config_info_ex, 'group'):
+                for group in config_info_ex.group:
+                    if not hasattr(group, 'vm'):
+                        continue
+                    for vm in group.vm:
+                        if vm.value == vm_ref.value:
+                            vm_groups.append(group.name)
+                            break
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: I still think you should only add a group to your list if it's an openstack-created group and not an admin-created group, i.e. if it's in `server_group_info`. I don't want you to delete my manually-assigned groups, I as a cluster operator need for maintenance on the cluster or whatever.

File: nova/virt/vmwareapi/cluster_util.py (hunk 9 above)
Comment (mmidolesov2) [reply nova/virt/vmwareapi/cluster_util.py]: When I ran the tests one of them was failing with config_spec referenced before assignment.

File: nova/virt/vmwareapi/vm_util.py (hunk 10 above)
Comment (mmidolesov2) [reply nova/virt/vmwareapi/vm_util.py]: Havent seen it Ill check

File: nova/virt/vmwareapi/cluster_util.py (hunk 9 above)
Comment (joker-official) [reply nova/virt/vmwareapi/cluster_util.py]: But this is not something that should happen in production, right? So shouldn't we fix the tests to look more like production instead?

Comment (mmidolesov2) [reply nova/virt/vmwareapi/cluster_util.py]: Actually we dont need it because we dont need a check there "cluster_config" will always has a value

File: nova/virt/vmwareapi/vmops.py (hunk 3 above)
Comment (joker-official) [reply nova/virt/vmwareapi/vmops.py]: This comment is still valid. We don't have to iterate further over the groups if we already found the one with the right name - doesn't matter if we took action on it or not.

Comment (joker-official) [reply nova/virt/vmwareapi/vmops.py]: This is still valid ...

**Comment by joker-official** [comment]
> ./nova/tests/unit/virt/vmwareapi/test_driver_api.py:1655:1: H401  docstring should not start with a space

File: nova/virt/vmwareapi/vmops.py [hunk 12]
```diff
@@ -1535,6 +1516,20 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
                 self._session._call_method(self._session.vim,
                                            "UnregisterVM", vm_ref)
                 LOG.debug("Unregistered the VM", instance=instance)
+
+                with lockutils.lock(server_group_info.uuid,
+                                lock_file_prefix='nova-vmware-server-group'):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Did you test this with a VM that's not in a server-group? Because I just did and the tests work. But if I add one line self.assertTrue(False)` to `nova.tests.unit.virt.vmwareapi.test_driver_api.VMwareAPIVMTestCase.test_destroy$` I can see the output the test was hiding. Nothing works as expected in there ... because that Exception is catched in the un-registering-error-handler ...

**Comment by joker-official** [comment]
Forgot to attach the log, which you can probably find on your own, but anyways:
```
    2019-11-27 12:29:02,964 INFO [nova.virt.vmwareapi.vm_util] Created folder: Instances in parent Folder-3.
    2019-11-27 12:29:02,968 ERROR [oslo_vmware.common.loopingcall] in fixed duration looping call
    Traceback (most recent call last):
      File "/nova/.tox/py27/local/lib/python2.7/site-packages/oslo_vmware/common/loopingcall.py", line 75, in _inner
        self.f(*self.args, **self.kw)
      File "/nova/.tox/py27/local/lib/python2.7/site-packages/oslo_vmware/api.py", line 452, in _poll_task
        raise task_ex
    FileNotFoundException: Error message
    2019-11-27 12:29:02,970 ERROR [oslo_vmware.common.loopingcall] in fixed duration looping call
    Traceback (most recent call last):
      File "/nova/.tox/py27/local/lib/python2.7/site-packages/oslo_vmware/common/loopingcall.py", line 75, in _inner
        self.f(*self.args, **self.kw)
      File "/nova/.tox/py27/local/lib/python2.7/site-packages/oslo_vmware/api.py", line 452, in _poll_task
        raise task_ex
    FileNotFoundException: Error message
    2019-11-27 12:29:02,971 ERROR [nova.tests.unit.virt.vmwareapi.fake] _retrieve_properties error
    Traceback (most recent call last):
      File "nova/tests/unit/virt/vmwareapi/fake.py", line 1588, in _retrieve_properties
        prop = Prop(prop_name, mdo.get(prop_name))
      File "nova/tests/unit/virt/vmwareapi/fake.py", line 230, in get
        return self.__getattr__(attr)
      File "nova/tests/unit/virt/vmwareapi/fake.py", line 256, in __getattr__
        'name': self.__class__.__name__})
    NovaException: Property summary not set for the managed object Datastore
    2019-11-27 12:29:02,972 WARNING [nova.virt.vmwareapi.vmops] Get esx cookies failed: 'NoneType' object has no attribute 'accessible'
    2019-11-27 12:29:02,972 ERROR [nova.tests.unit.virt.vmwareapi.fake] _retrieve_properties error
    Traceback (most recent call last):
      File "nova/tests/unit/virt/vmwareapi/fake.py", line 1576, in _retrieve_properties
        path = obj.selectSet[0].path
    TypeError: 'DataObject' object does not support indexing
    2019-11-27 12:29:02,981 WARNING [nova.virt.vmwareapi.vmops] In vmwareapi:vmops:_destroy_instance, got this exception while un-registering the VM: 'NoneType' object has no attribute 'uuid'
    2019-11-27 12:29:02,981 ERROR [nova.tests.unit.virt.vmwareapi.fake] _retrieve_properties error
    Traceback (most recent call last):
      File "nova/tests/unit/virt/vmwareapi/fake.py", line 1585, in _retrieve_properties
        mdo = _db_content[spec_type][mdo_ref]
    KeyError: <nova.tests.unit.virt.vmwareapi.fake.ManagedObjectReference object at 0x7faa98ea72d0>
```

**Comment by joker-official** [comment]
Please also add some logging around the vmgroup-removal. Since it's a vcenter task, it might take some time and we need to know what nova-compute does in debug-level at least.

**Comment by mmidolesov2** [comment]
Changes are added.

File: nova/virt/vmwareapi/vm_util.py [hunk 13]
```diff
@@ -1315,7 +1315,7 @@ def update_cluster_placement(session, context, instance, cluster, vm_ref):
         return
 
     with lockutils.lock(server_group_info.uuid,
-                        lock_file_prefix='nova-vmware-server-group'):
+                        lock_file_prefix='vmware-vm-group-policy'):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vm_util.py]: I don't think that lock is necessary, since `cluster_util.update_placement()` is guarded by `@utils.synchronized` already.

File: nova/virt/vmwareapi/vmops.py [hunk 14]
```diff
@@ -1517,18 +1517,24 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
+                                              server_group_info.uuid)
+                                    cluster_util.delete_vm_group(
+                                        self._session, self._cluster,
+                                        group)
+                                    LOG.debug("VM group %s deleted"
+                                              " successfully" % group.name)
+                                break
 
             except Exception as excep:
                 LOG.warning("In vmwareapi:vmops:_destroy_instance, got "
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: This log-message is not correct anymore, if the server-group deletion fails. Please use an own try-except for the new code.

File: nova/virt/vmwareapi/vmops.py [hunk 15]
```diff
@@ -1517,18 +1517,24 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
-                    for group in cluster_config.group:
-                        if group.name == server_group_info.uuid:
-                            if not hasattr(group, 'vm'):
-                                cluster_util.delete_vm_group(
-                                    self._session, self._cluster,
-                                    group)
-                            break
+                if server_group_info:
+                    with lockutils.lock(server_group_info.uuid,
+                                    lock_file_prefix='vmware-vm-group-policy'):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: I think this should use the same lock as `cluster_util.update_placement()` as it changes the same thing. Or do you see a need for a group-specific lock?

File: nova/virt/vmwareapi/vm_util.py (hunk 13 above)
Comment (mmidolesov2) [reply nova/virt/vmwareapi/vm_util.py]: Yes since both end up doing reconfigureCluster this is not needed actually. I'll remove it.

File: nova/virt/vmwareapi/vmops.py (hunk 15 above)
Comment (mmidolesov2) [reply nova/virt/vmwareapi/vmops.py]: Yeah since I removed the lock in vm_util now we should use the same lock name. So I changed it.

File: nova/virt/vmwareapi/vmops.py [hunk 16]
```diff
@@ -1535,6 +1516,30 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
                 self._session._call_method(self._session.vim,
                                            "UnregisterVM", vm_ref)
                 LOG.debug("Unregistered the VM", instance=instance)
+
+                if server_group_info:
+                    with lockutils.lock('vmware-vm-group-policy'):
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: You need to set the same `lock_file_prefix` as `cluster_util.update_placement()` uses. Please look at `nova.utils` for the `@synchronized` decorator.

File: nova/virt/vmwareapi/vm_util.py [hunk 17]
```diff
@@ -708,6 +709,13 @@ def _get_device_disk_type(device):
             return constants.DEFAULT_DISK_TYPE
 
 
+def get_cluster_property(session, prop_list, cluster):
+    cluster_config = session._call_method(
+        vutil, "get_object_property", cluster,
+        prop_list)
+
+    return cluster_config
```
Comment (joker-official) [inline nova/virt/vmwareapi/vm_util.py]: This function is now duplicated to `cluster_util`. I think you should remove it here.

File: nova/virt/vmwareapi/vm_util.py [hunk 18]
```diff
@@ -22,6 +22,7 @@
 import copy
 import functools
 
+from oslo_concurrency import lockutils
```
Comment (joker-official) [inline nova/virt/vmwareapi/vm_util.py]: This import seems unused. Seeing that, I know you haven't run `tox -e pep8`. Please update and tell me, when you have run it.

File: nova/virt/vmwareapi/vmops.py [hunk 19]
```diff
@@ -1537,18 +1538,16 @@ def _destroy_instance(self, context, instance, destroy_disks=True):
...
-                                            config_info_ex.group[key])
-                                break
-                        break
+            if hasattr(config_info_ex, 'group'):
+                for group in config_info_ex.group:
+                    if not hasattr(group, 'vm'):
+                        continue
+                    for vm in group.vm:
+                        if vm.value == vm_ref.value:
+                            vm_groups.append(group.name)
```
Comment (joker-official) [inline nova/virt/vmwareapi/vmops.py]: Is this unused now that we do an updated query down at the end?

Comment (mmidolesov2) [reply nova/virt/vmwareapi/vmops.py]: Yes I'll remove what needed.

**Comment by joker-official** [comment]
Superseded by https://github.com/sapcc/nova/pull/146
